packages/cojson-transport-ws/src/index.ts

```diff
@@ -1,108 +1,121 @@
-                    console.error(
-                        "Error while trying to close ws on ping timeout",
-                        e,
-                    );
-                }
-            }, 2500);
+        const isSyncMessage = (msg: unknown): msg is SyncMessage => {
+            // boo side-effects
 
-            if (msg.type === "ping") {
-                // eslint-disable-next-line @typescript-eslint/no-explicit-any
-                (globalThis as any).jazzPings =
-                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-                    (globalThis as any).jazzPings || [];
-                // eslint-disable-next-line @typescript-eslint/no-explicit-any
-                (globalThis as any).jazzPings.push({
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            if ((msg as any)?.type === "ping") {
+                const ping = msg as PingMsg;
+                g.jazzPings ||= [];
+                g.jazzPings.push({
                     received: Date.now(),
-                    sent: msg.time,
-                    dc: msg.dc,
+                    sent: ping.time,
+                    dc: ping.dc,
                 });
-                return;
-            } else {
-                void Effect.runPromise(
-                    Queue.offer(incoming, Either.right(msg)),
-                );
+                return false;
             }
-        });
+            return true;
+        };
 
-        ws.addEventListener("open", () => {
-            void Stream.fromQueue(outgoing).pipe(
-                Stream.runForEach((msg) =>
-                    Effect.sync(() => ws.send(JSON.stringify(msg))),
+        yield* Queue.take(outgoing).pipe(
+            Effect.andThen((message) => ws.send(JSON.stringify(message))),
+            Effect.forever,
+            Effect.whenEffect(
+                fromEventListener(ws, "open").pipe(Effect.as(true)),
+            ),
+            Effect.forkDaemon, // :/
+        );
+
+        type E = WebsocketEvents["message"];
+        const messages = Stream.fromEventListener<E>(ws_, "message").pipe(
+            Stream.timeoutFail(() => new PingTimeoutError(), "2.5 seconds"),
+            Stream.tapError((_e) =>
+                Console.warn("Ping timeout").pipe(
+                    Effect.zipRight(Effect.try(() => ws.close())),
+                    Effect.catchAll((e) =>
+                        Console.error(
+                            "Error while trying to close ws on ping timeout",
+                            e,
+                        ),
+                    ),
                 ),
-                Effect.runPromise,
-            );
-        });
+            ),
+            Stream.mergeLeft(closed),
+            Stream.map((_) => JSON.parse(_.data as string)),
+            Stream.filter(isSyncMessage),
+            Stream.buffer({ capacity: "unbounded" }),
+            Stream.onDone(() => Queue.shutdown(outgoing)),
+        );
```

Comment:
creates a `Stream` `fromEventListener`, it must emit *any* value **at most** 2.5 seconds apart or we `timeout` and fail. If we do, we print a warning and `try` to close the socket. If it doesn't work, we `catch` the `UnknownError` and report it to the console. 

Then `merge` it with the `closed` stream, ignore anything that comes out of that stream--- here I'm relying on the error handling logic of effect. `closed` is a 0-element `Stream` that if it manages to evaluate, meaning that the `close` event was emitted, it would just immediately fail with `DisconnectedError`. Practically speaking, this error and the fancy way to create it is only required because we can't use scopes.

We then `filter` out all the pings (and save them on the globalThis as a side effect) and put everything into an unlimited `buffer`. This is a bad practice. But its the same thing we had before. Important to note, I've added a behavior change here (everything else works logically exactly the same). The previous impl would listen for messages from t0 and until the stream was closed by the consumer or the socket was closed by anything and there was a lot of hope going around. This impl only listens for messages once the `incoming` stream is evaluated. That can be in t0, t1, or any tn. I make the assumption that it would not be any later than t1 and if thats the case the behavior should be entirely non-breaking. The tests pass too.

Regardless, once this stream is evaluated once and is done, we shut down the `outgoing` queue so the consumer won't attempt to offer data that will never be read.

Reply:
The tests don't test this package, but I verified manually that the chat example works <picture data-single-emoji=":meow_verified:" title=":meow_verified:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/meow_verified/e5e955816b46e3ce.png" alt=":meow_verified:" align="absmiddle"></picture>

Reply:
The chat example works perfectly from what I could tell 😄